<a href="https://colab.research.google.com/github/google/applied-machine-learning-intensive/blob/master/content/02_data/05_exploratory_data_analysis/colab-part1.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#### Copyright 2020 Google LLC.

In [ ]:
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

# Exploratory Data Analysis

[Exploratory Data Analysis](https://en.wikipedia.org/wiki/Exploratory_data_analysis), often shortened to EDA, is a term that you'll hear quite a bit in the field of data science. EDA is the process of examining a dataset to find facts about the data and communicating those facts, often through visualizations.

In order to explore the data and visualize it, some modifications might need to be made to the data along the way. This is often referred to as *data preprocessing*. Though data preprocessing is technically different from EDA, EDA often exposes problems with the data that need to be fixed in order to continue exploring. Because of this tight coupling, we'll clean the data as necessary to help understand the data.

In this lab we will apply our Pandas knowledge to explore a dataset about chocolate. Part 1 of the lab will explore each column in our dataset individually. Part 2 will take the results of our preprocessed data and search for patterns across columns and rows.

## Introduction



### The Dataset: Chocolate Bar Ratings

In this lab we will use a [chocolate bar ratings dataset](https://www.kaggle.com/rtatman/chocolate-bar-ratings). The dataset is from the [Flavors of Cacao](http://flavorsofcacao.com/flavor.html) data.

On the [Kaggle page for the dataset](https://www.kaggle.com/rtatman/chocolate-bar-ratings), we can find some basic information about the dataset. For instance, there are over 1,700 chocolate bars that have been rated. We can also preview the columns found in the dataset:

Column | Data Type | Description
-------|-----------|-------------
Company (Maker-if known) | String | Name of the company manufacturing the bar.
Specific Bean Origin or Bar Name | String | The specific geo-region of origin for the bar.
REF | Number | A value linked to when the review was entered in the database. Higher = more recent.
Review Date | Number | Date of publication of the review.
Cocoa Percent | String | Cocoa percentage (darkness) of the chocolate bar being reviewed.
Company Location | String | Manufacturer base country.
Rating | Number | Expert rating for the bar.
BeanType | String | The variety (breed) of bean used, if provided.
Broad Bean Origin | String | The broad geo-region of origin for the bean.

This is an interesting dataset. Think of the questions that you might be able to answer! A few could be:

*   Is there a relationship between numeric rating and properties such as percentage of cocoa, bean type, origin, and maker?
*   Are some of the properties of cacao beans correlated?
*   Where are the top chocolate bars from?
*   Are there multiple entries for the same bar from the same maker, but with different ratings over the years? If so, has there been any change in the chocolate bar that could account for the differences?
*   Do makers who produce a wide variety of bars have a higher chance of creating a top-rated chocolate bar?

I'm sure you can think of even more. So, what are we waiting for? Let's load the data!

## Acquiring the Data

The data is hosted on Kaggle, so we can use our Kaggle credentials to download the data into the lab. The dataset is located at [https://www.kaggle.com/rtatman/chocolate-bar-ratings](https://www.kaggle.com/rtatman/chocolate-bar-ratings). We can use the `kaggle` command line utility to do this.

First off, upload your `kaggle.json` file into the lab now.

Next, run the following command to get the credential files set to the right permissions and located in the correct spot.

In [ ]:
! chmod 600 kaggle.json && (ls ~/.kaggle 2>/dev/null || mkdir ~/.kaggle) && mv kaggle.json ~/.kaggle/ && echo 'Done'

Now we can run the `kaggle` command to actually download the data.

In [ ]:
! kaggle datasets download rtatman/chocolate-bar-ratings
! ls

We now have our data downloaded to our virtual machine and stored in the file `chocolate-bar-ratings.zip`.

## Creating a `DataFrame`

We now need to load the data into memory. We can do this easily using Pandas' `read_csv()` function.

In [1]:
import pandas as pd

df = pd.read_csv('chocolate-bar-ratings.zip')
df

Company \n(Maker-if known) Specific Bean Origin\nor Bar Name   REF  \
0                      A. Morin                       Agua Grande  1876   
1                      A. Morin                             Kpime  1676   
2                      A. Morin                            Atsane  1676   
3                      A. Morin                             Akata  1680   
4                      A. Morin                            Quilla  1704   
...                         ...                               ...   ...   
1790                     Zotter                              Peru   647   
1791                     Zotter                             Congo   749   
1792                     Zotter                      Kerala State   749   
1793                     Zotter                      Kerala State   781   
1794                     Zotter                Brazil, Mitzi Blue   486   

      Review\nDate Cocoa\nPercent Company\nLocation  Rating Bean\nType  \
0             2016            63%            France    3.75              
1             2015            70%            France    2.75              
2             2015            70%            France    3.00              
3             2015            70%            France    3.50              
4             2015            70%            France    3.50              
...            ...            ...               ...     ...        ...   
1790          2011            70%           Austria    3.75              
1791          2011            65%           Austria    3.00  Forastero   
1792          2011            65%           Austria    3.50  Forastero   
1793          2011            62%           Austria    3.25              
1794          2010            65%           Austria    3.00              

     Broad Bean\nOrigin  
0              Sao Tome  
1                  Togo  
2                  Togo  
3                  Togo  
4                  Peru  
...                 ...  
1790               Peru  
1791              Congo  
1792              India  
1793              India  
1794             Brazil  

[1795 rows x 9 columns]

Let's also make sure that our data types match what was documented:

In [2]:
df.dtypes

Company \n(Maker-if known)            object
Specific Bean Origin\nor Bar Name     object
REF                                    int64
Review\nDate                           int64
Cocoa\nPercent                        object
Company\nLocation                     object
Rating                               float64
Bean\nType                            object
Broad Bean\nOrigin                    object
dtype: object

In this output, `object` types are strings while `int64` types are whole numbers and `float64` types are fractional numbers. This seems to match the documentation that we saw for the dataset.

From just a glance at the `DataFrame`, we can see a few facts about our data:

* There are 1,795 rows and 9 columns.
* The columns are the columns we expected based on the documentation, though some have `\n` (new line) embedded in them. We'll need to clean that up.
* The data seems to be sorted by the 'Company' column.
* There is definitely some missing data, as we can see in the 'Bean Type' column.

We will look more closely at each column throughout this lab.

## Cleaning Up Column Names

One of the more frustrating aspects of this dataset is the poor format of the column names. Typing 'Specific Bean Origin\nor Bar Name' in order to access the column is painful.

So our first order of business will be to update the column names.

In [3]:
df.columns = [
  'Company',
  'Specific Bean Origin',
  'REF',
  'Review Date',
  'Cocoa Percent',
  'Company Location',
  'Rating',
  'Bean Type',
  'Broad Bean Origin'
]

df

Company Specific Bean Origin   REF  Review Date Cocoa Percent  \
0     A. Morin          Agua Grande  1876         2016           63%   
1     A. Morin                Kpime  1676         2015           70%   
2     A. Morin               Atsane  1676         2015           70%   
3     A. Morin                Akata  1680         2015           70%   
4     A. Morin               Quilla  1704         2015           70%   
...        ...                  ...   ...          ...           ...   
1790    Zotter                 Peru   647         2011           70%   
1791    Zotter                Congo   749         2011           65%   
1792    Zotter         Kerala State   749         2011           65%   
1793    Zotter         Kerala State   781         2011           62%   
1794    Zotter   Brazil, Mitzi Blue   486         2010           65%   

     Company Location  Rating  Bean Type Broad Bean Origin  
0              France    3.75                     Sao Tome  
1              France    2.75                         Togo  
2              France    3.00                         Togo  
3              France    3.50                         Togo  
4              France    3.50                         Peru  
...               ...     ...        ...               ...  
1790          Austria    3.75                         Peru  
1791          Austria    3.00  Forastero             Congo  
1792          Austria    3.50  Forastero             India  
1793          Austria    3.25                        India  
1794          Austria    3.00                       Brazil  

[1795 rows x 9 columns]

That's much better, but the columns are also in an odd order. Information about the company is spread across the columns, and so is the information about the cacao bean. Let's order the columns a little more meaningfully.

This order makes a little more sense:

**Company Information:**
* Company
* Company Location

**Chocolate Bar Information**
* Bean Type
* Specific Bean Origin
* Broad Bean Origin
* Cocoa Percent

**Review Information**
* REF
* Review Date
* Rating

We can reorder the columns by specifically selecting the columns in order and reassigning them to the `df` variable:

In [4]:
df = df[[
  'Company',
  'Company Location',
  'Bean Type',
  'Specific Bean Origin',
  'Broad Bean Origin',
  'Cocoa Percent',
  'REF',
  'Review Date',
  'Rating',
]]

df

Company Company Location  Bean Type Specific Bean Origin  \
0     A. Morin           France                     Agua Grande   
1     A. Morin           France                           Kpime   
2     A. Morin           France                          Atsane   
3     A. Morin           France                           Akata   
4     A. Morin           France                          Quilla   
...        ...              ...        ...                  ...   
1790    Zotter          Austria                            Peru   
1791    Zotter          Austria  Forastero                Congo   
1792    Zotter          Austria  Forastero         Kerala State   
1793    Zotter          Austria                    Kerala State   
1794    Zotter          Austria              Brazil, Mitzi Blue   

     Broad Bean Origin Cocoa Percent   REF  Review Date  Rating  
0             Sao Tome           63%  1876         2016    3.75  
1                 Togo           70%  1676         2015    2.75  
2                 Togo           70%  1676         2015    3.00  
3                 Togo           70%  1680         2015    3.50  
4                 Peru           70%  1704         2015    3.50  
...                ...           ...   ...          ...     ...  
1790              Peru           70%   647         2011    3.75  
1791             Congo           65%   749         2011    3.00  
1792             India           65%   749         2011    3.50  
1793             India           62%   781         2011    3.25  
1794            Brazil           65%   486         2010    3.00  

[1795 rows x 9 columns]

## Examining Each Column


In this section we will examine each column to learn about the data in the column. We will also make changes to the data as needed.

### Column: Company

The 'Company' column is the first in the list, so let's look at it first.

We can tell that the column contains string values. Let's see if any are missing:

In [ ]:
df['Company'].isnull().any()

No data is missing. Let's now see how many distinct values there are:

In [ ]:
df['Company'].unique().size

A few hundred is not a terribly long list. Let's print the list in alphabetical order to see how it looks.

In [ ]:
for company in sorted(df['Company'].unique()):
  print(company)

This is some interesting data. Looking at it raises many questions. For instance:

* Should company names like 'Vintage Plantations' and 'Vintage Plantations (Tulicorp)' be changed to the same name?
* Is 'Cacao de Origin' a misspelling of 'Cacao de Origen'?
* Is 'Shattel' a misspelling of 'Shattell'?

These are the types of things you'll see and questions you'll ask when you encounter a new dataset. Rarely is the data in perfect condition. Often you'll spend a considerable amount of time researching topics related to the data in order to make a call about repairing aspects of the data.

In this particular case, it would be great if we could find a master list of all of the chocolate makers in the world. We could then cross reference the names in the dataset with the names in the master list.

Unfortunately, we don't have a master list of chocolate makers. Instead, we will have to rely on manually inspecting the data and researching when things don't look right.

Let's say that for now we are confident that 'Cacao de Origin' and 'Shattel' are misspellings, so we will correct that data. We aren't confident enough to change any of the names with parentheses in them though.

Let's fix our misspellings!

#### Exercise 1: Fixing Misspellings

We have decided that we would like to change every instance of 'Cacao de Origin' to 'Cacao de Origen' and every instance of 'Shattel' to 'Shattell' in the 'Company' column of our dataset. Write the code to modify the values. Make sure your code doesn't have any warnings. At the end of the code block, print the number of unique company names when you are done. There should be two less columns than what you saw above.

**Student Solution**

In [7]:
import pandas as pd
import numpy as np

df = pd.read_csv('chocolate-bar-ratings.zip')
df.columns = ['Company', 'Specific Bean Origin', 'REF', 'Review Date',
              'Cocoa Percent', 'Company Location', 'Rating', 'Bean Type',
              'Broad Bean Origin']
df = df[['Company', 'Company Location', 'Bean Type', 'Specific Bean Origin',
         'Broad Bean Origin', 'Cocoa Percent', 'REF', 'Review Date', 'Rating']]

# Change 'Shattel' to 'Shattell'
df['Company'] = np.where(df['Company'] == 'Shattel', 'Shattell', df['Company'])

# Change 'Cacao de Origin' to 'Cacao de Origen'
df['Company'] = np.where(df['Company'] == 'Cacao de Origin', 'Cacao de Origen', df['Company'])

# Print the number of unique company names
df['Company'].nunique()

414

---

### Column: Company Location

The [documentation](https://www.kaggle.com/rtatman/chocolate-bar-ratings) describes the 'Company Location' column as "*Manufacturer base country*."

Let's take a look at the data. As always, we'll first check to see if any data is missing.

In [ ]:
df['Company Location'].isna().any()

No missing data.

Now we can see how many unique values there are:

In [ ]:
df['Company Location'].unique().shape

There are just 60 locations, which is small enough that we can manually inspect the values. Let's print the data.

In [ ]:
for location in sorted(df['Company Location'].unique()):
  print(location)

Overall, the data looks pretty clean. The column is supposed to contain countries and *most* entries are countries. There are a few problems with the country data though. We found at least five errors in the data. Let's see what you can find.

#### Exercise 2: Fixing Company Location Data

There are at least five errors in the company location data that need to be fixed. Some are fairly easy to spot (spelling errors), but some do require knowledge of what constitutes a country. Take some time to look at the data, and see if you can spot at least two of the issues. Write code to fix the issues.

**Student Solution**

In [8]:
# Fix at least two issues with the 'Company Location' data
df['Company Location'] = np.where(df['Company Location'] == 'Nicragua', 'Nicaragua', df['Company Location'])
df['Company Location'] = np.where(df['Company Location'] == 'Domincan Republic', 'Dominican Republic', df['Company Location'])

---

### Column: Bean Type

Now that our company data is looking a little better, let's move into data about the cocoa going into the chocolate bar itself. The first piece of data is the 'Bean Type'. 'Bean Type' is defined as "*The variety (breed) of bean used, if provided*". This hints that there will be some missing data. Let's check and see.

In [ ]:
df['Bean Type'].isna().any()

Indeed, we have missing data. Let's see how much is missing.

In [ ]:
df[df['Bean Type'].isna()].count()

Only one row of data is missing 'Bean Type'. Let's take a look at that row.

- MNAR: Not missing at random.
- MAR: Missing at random.
- MCAR: Completely missing at random

In [9]:
df[df['Bean Type'].isna()]

Company Company Location Bean Type  \
1544    Soma           Canada       NaN   

                         Specific Bean Origin   Broad Bean Origin  \
1544  Three Amigos(Chuao, Wild Bolivia, D.R.)  Ven, Bolivia, D.R.   

     Cocoa Percent  REF  Review Date  Rating  
1544           70%  676         2011     4.0

Now we have a choice to make about how to handle this missing data. Some options include:

* Leave it as is
* Remove the entire row
* Fill in the data with some value

Leaving undefined values lying around in our data can be problematic. Missing values are not counted and can be tricky to program around.

Removing the entire row actually isn't a bad option in this case. Since it is only one row out of over 1,700, it likely won't have too much effect on any analysis that we do.

As for filling in the row, we can:

* Use 'Unknown' or some other placeholder value
* Actually do research to find the true missing value
* See if there is a reasonable value already in the data

In this case, we are just going to replace the missing value with 'Unknown'.

In [10]:
df.loc[df['Bean Type'].isna(), 'Bean Type'] = 'Unknown'
df[df['Bean Type'].isna()]

Empty DataFrame
Columns: [Company, Company Location, Bean Type, Specific Bean Origin, Broad Bean Origin, Cocoa Percent, REF, Review Date, Rating]
Index: []

Now we can see how many unique bean types we have.

In [11]:
df['Bean Type'].unique().size

42

Only 42, let's print them out.

In [12]:
for t in sorted(df['Bean Type'].unique()):
  print(t)

Amazon
Amazon mix
Amazon, ICS
Beniano
Blend
Blend-Forastero,Criollo
CCN51
Criollo
Criollo (Amarru)
Criollo (Ocumare 61)
Criollo (Ocumare 67)
Criollo (Ocumare 77)
Criollo (Ocumare)
Criollo (Porcelana)
Criollo (Wild)
Criollo, +
Criollo, Forastero
Criollo, Trinitario
EET
Forastero
Forastero (Amelonado)
Forastero (Arriba)
Forastero (Arriba) ASS
Forastero (Arriba) ASSS
Forastero (Catongo)
Forastero (Nacional)
Forastero (Parazinho)
Forastero(Arriba, CCN)
Forastero, Trinitario
Matina
Nacional
Nacional (Arriba)
Trinitario
Trinitario (85% Criollo)
Trinitario (Amelonado)
Trinitario (Scavina)
Trinitario, Criollo
Trinitario, Forastero
Trinitario, Nacional
Trinitario, TCGA
Unknown
 


The data looks pretty good. But there is a small little problem. After 'Unknown' there seems to be an empty line. What is that?

It turns out that it is a whitespace character. We thought we had only one missing value, but it looks like there are some values that are present but are white space. Let's see how many.

White space can be tricky because there are many different encodings that render as white space. Let's find out exactly which space character this is.

To get the space(s) we can sort the 'Bean Type' values again and get the last one, since we see the space last in the list. We can then print the space as hexadecimal characters.

In [13]:
space = sorted(df['Bean Type'].unique())[-1]
print(", ".join("0x{:02x}".format(ord(c)) for c in space))

0xa0


We get `0xa0` which is the ASCII code for [non-breaking space](https://en.wikipedia.org/wiki/Non-breaking_space). This is different from the white space that you get when you hit the space bar. That space is encoded `0x20`.

Let's see how many of these there are:

In [14]:
df[df['Bean Type'] == chr(0xa0)]

Company Company Location Bean Type Specific Bean Origin  \
0     A. Morin           France                    Agua Grande   
1     A. Morin           France                          Kpime   
2     A. Morin           France                         Atsane   
3     A. Morin           France                          Akata   
4     A. Morin           France                         Quilla   
...        ...              ...       ...                  ...   
1787    Zotter          Austria                  Santo Domingo   
1789    Zotter          Austria                  Indianer, Raw   
1790    Zotter          Austria                           Peru   
1793    Zotter          Austria                   Kerala State   
1794    Zotter          Austria             Brazil, Mitzi Blue   

       Broad Bean Origin Cocoa Percent   REF  Review Date  Rating  
0               Sao Tome           63%  1876         2016    3.75  
1                   Togo           70%  1676         2015    2.75  
2                   Togo           70%  1676         2015    3.00  
3                   Togo           70%  1680         2015    3.50  
4                   Peru           70%  1704         2015    3.50  
...                  ...           ...   ...          ...     ...  
1787  Dominican Republic           70%   879         2012    3.75  
1789                               58%   883         2012    3.50  
1790                Peru           70%   647         2011    3.75  
1793               India           62%   781         2011    3.25  
1794              Brazil           65%   486         2010    3.00  

[887 rows x 9 columns]

Almost 900! Let's encode those as 'Unknown' also.

#### Exercise 3: Fixing Non-Breaking Space

There are non-breaking space characters, `0xa0` in the 'Bean Type' column. Replace these values with the word 'Unknown'.

**Student Solution**

In [16]:
# Your Code Goes Here
df['Bean Type'] = np.where(df['Bean Type'] == chr(0xa0), 'Unknown', df['Bean Type'])

---

### Column: Specific Bean Origin

Let's look at our next column: 'Specific Bean Origin'. 'Specific Bean Origin' is a string column that contains the "*specific geo-region of origin for the bar*."

First, we'll see if we are missing any data in the 'Specific Bean Origin' column.

In [17]:
df['Specific Bean Origin'].isna().any()

False

Good, we don't have any 'N/A' data. But we learned from the 'Bean' column that we also need to check string columns for being only white space.

A good way to do this is to apply a function that strips leading and trailing white space from every value in a column, and see if the resulting string is zero-length.

In [18]:
df[df['Specific Bean Origin'].apply(lambda x: x.strip()).str.len() == 0]

Empty DataFrame
Columns: [Company, Company Location, Bean Type, Specific Bean Origin, Broad Bean Origin, Cocoa Percent, REF, Review Date, Rating]
Index: []

Here we can see that no data was returned, so we don't have any 'Specific Bean Origin' values that are only spaces.

If you run this function and get an error about numbers/floats not having a strip function, you likely have N/A values in your column. Always check `isna()` first.

Now that we know that every row has a 'Specific Bean Origin' value, let's see how many unique values we have.

In [ ]:
df['Specific Bean Origin'].unique().size

Over 1,000 values! That is quite a bit of data to manually sift through. Let's look at the first bit of data, up until the first origin that starts with 'B'.

In [ ]:
for origin in sorted(df['Specific Bean Origin'].unique()):
  if origin.startswith('B'):
    break
  print(origin)

This is some pretty ugly data. Most (but not all) rows contain the bean's geographical origin, but some seem to include the year and/or batch numbers as well, and some seem to contain different information entirely ("100 percent").

Looking at the data, we can also see some things that look odd. For instance, "Akesson Estate" and "Akesson's Estate" are likely the same origin. Also, "Ambolikapkly P." clearly looks like a misspelling of "Ambolikapiky P."

We could make all of the "Akesson" origins look the same, but should we? First, let's look at the entire rows for the offending data.

In [ ]:
df[(df['Specific Bean Origin'] == 'Akesson Estate') | \
   (df['Specific Bean Origin'] == "Akesson's Estate")]

It is interesting that all of the bean types and origins are alike. It looks like Akesson('s) Estate serves many companies though.

It is tempting to go ahead and change the "Specific Bean Origin" values to make them match, but it is better to do more research into the industry before making those sorts of changes. You might disagree with this decision, and that is perfectly fine. When working with datasets, you will often have to make difficult calls to deal with ambiguous data. Different people will make different decisions, and that's okay.

The "Ambolikapkly P." issue is a little more obvious and can be validated with a quick internet search. The "Ambolikapkly" spelling shows up very few times and always in the context of this data set. The other spelling is much more common. Let's go ahead and fix that.

In [ ]:
df.loc[df['Specific Bean Origin'] == 'Ambolikapkly P.', 
       'Specific Bean Origin'] = 'Ambolikapiky P.'

#### Exercise 4: Finding and Repairing Bad Data

There are a few more obvious errors in the 'Specific Bean Origin' column of the dataset. Print out the column, scan the output, and see if you can find any more errors. Write the code to fix the errors. Find at least one error to fix.

The code to print the dataset is below.

In [19]:
for origin in sorted(df['Specific Bean Origin'].unique()):
  print(origin)

"heirloom", Arriba Nacional
100 percent
2009 Hapa Nibby
A case of the Xerces Blues, triple roast
ABOCFA Coop
Abinao
Abstract S. w/ Jamaica nibs,batch abs60323.0
Acarigua, w/ nibs
Acopagro
Acul-du-Nord, 2015
Africa
Africa meets Latina
AgroCriso Plantation
Agua Fria; Sucre region
Agua Grande
Akata
Akesson Estate
Akesson P.
Akesson's E., Sambirano V.
Akesson's Estate
Akesson's, batch 4411
Akessons Estate, Sambirano, 2013
Akessons Estate, Sambirano, Ambanja
Akosombo
Almendra Blanca, batch 1004
Alpaco
Alta Verapaz, 2014
Alto Beni
Alto Beni, Covendo Region
Alto Beni, Cru Savage
Alto Beni, Palos Blanco
Alto Beni, Palos Blancos
Alto Beni, Upper Rio Beni, 2014
Alto Beni, Upper Rio Beni, 2015
Alto Beni, Wild Bolivian
Alto Beni, Wild Harvest, Itenez R. 24hr c.
Alto Beni, Wild Harvest, Itenez R., 60hr c.
Alto Beni, Wild Harvest, Limited Ed.
Amazon Basin Blend
Amazonas
Amazonas Frucht
Amazonas w/ nibs
Amazonia
Ambanja, Sambirano Valley
Ambanja, Tsara Valley
Ambanja, batch 1 SRB
Ambolikapiky
Ambolik

**Student Solution**

In [20]:
# Repair the data
df['Specific Bean Origin'] = np.where(df['Specific Bean Origin'] == 'Bolivian', 'Bolivia', df['Specific Bean Origin'])

---

#### Exercise 5: Top Specific Bean Origins

There are just over 1,000 unique specific bean origins and over 1,700 entries in the dataset. Write code to find the top five most repeated origins. Print the origins and the number of times that each appears in the dataset.

**Student Solution**

In [23]:
# Find the top 5 bar origins
df.groupby('Specific Bean Origin')['Specific Bean Origin'].count().sort_values(ascending=False).head(n=5)

Specific Bean Origin
Madagascar            57
Peru                  45
Ecuador               42
Dominican Republic    37
Venezuela             21
Name: Specific Bean Origin, dtype: int64

---

### Column: Broad Bean Origin

The 'Broad Bean Origin' is the "*broad geo-region of origin for the bean.*" In theory, this should be broader regions than the 'Specific Bean Origin' that we just worked with.

Let's dive in. First things first, let's check for N/A values.

In [24]:
df[df['Broad Bean Origin'].isna()].count()

Company                 1
Company Location        1
Bean Type               1
Specific Bean Origin    1
Broad Bean Origin       0
Cocoa Percent           1
REF                     1
Review Date             1
Rating                  1
dtype: int64

It looks like we are missing one origin. Let's take a look at the record.

In [25]:
df[df['Broad Bean Origin'].isna()]

Company Company Location   Bean Type Specific Bean Origin  \
1072  Mast Brothers           U.S.A.  Trinitario           Madagascar   

     Broad Bean Origin Cocoa Percent  REF  Review Date  Rating  
1072               NaN           72%  999         2012     2.5

The one record has a 'Specific Bean Origin' of 'Madagascar'. Let's see if there are any other chocolates from that same specific origin.

In [ ]:
df[df['Specific Bean Origin'] == 'Madagascar']

Quite a few! And they all have a 'Broad Bean Origin' of 'Madagascar', except for our one missing value. It is probably safe to just set the missing value to 'Madagascar' also.

In [28]:
df.loc[(df['Specific Bean Origin'] == 'Madagascar') &
       (df['Broad Bean Origin'].isna()),
       'Broad Bean Origin'] = 'Madagascar'

df[df['Broad Bean Origin'].isna()]

Empty DataFrame
Columns: [Company, Company Location, Bean Type, Specific Bean Origin, Broad Bean Origin, Cocoa Percent, REF, Review Date, Rating]
Index: []

Now that we have all of the N/A values handled, let's see if we have an issue with spaces.

In [29]:
df[df['Broad Bean Origin'].apply(lambda x: x.strip()).str.len() == 0]

Company Company Location Bean Type  Specific Bean Origin  \
77         Amedei            Italy     Blend                  Nine   
85         Amedei            Italy     Blend         Toscano Black   
86         Amedei            Italy     Blend         Toscano Black   
87         Amedei            Italy     Blend         Toscano Black   
144   Bahen & Co.        Australia   Unknown            Houseblend   
...           ...              ...       ...                   ...   
1766        Zak's           U.S.A.   Unknown  House Blend, Batch 2   
1774       Zokoko        Australia   Unknown         Goddess Blend   
1778       Zotter          Austria   Unknown                   Raw   
1780       Zotter          Austria   Unknown       Amazonas Frucht   
1789       Zotter          Austria   Unknown         Indianer, Raw   

     Broad Bean Origin Cocoa Percent   REF  Review Date  Rating  
77                               75%   111         2007    4.00  
85                               63%   170         2007    3.50  
86                               70%    40         2006    5.00  
87                               66%    75         2006    4.00  
144                              80%  1474         2015    3.25  
...                ...           ...   ...          ...     ...  
1766                             60%  1582         2015    3.00  
1774                             65%  1780         2016    3.25  
1778                             80%  1205         2014    2.75  
1780                             65%   801         2012    3.50  
1789                             58%   883         2012    3.50  

[73 rows x 9 columns]

There are spaces in 73 rows of the data. Let's see what those space values are.

In [30]:
spaces_df = df[df['Broad Bean Origin'].apply(
    lambda x: x.strip()).str.len() == 0]

for space in spaces_df['Broad Bean Origin'].unique():
  print(", ".join("0x{:02x}".format(ord(c)) for c in space))

0xa0


It is that pesky `0xa0` again.

We can fix this by replacing all of the `0xa0` values with 'Unknown'. However, an even better fix would be if we could find similar chocolates with the same 'Specific Bean Origin' and then derive the 'Broad Bean Origin' from that.

Let's see if it is even possible. To do that we can find all of the 'Specific Bean Origin' values for rows with 'Broad Bean Origin' and those without. Then we can use `pd.merge()` to combine the two. If you remember, `pd.merge()` returns only the values which appear in both of the given Series. This means that the return value will show us which values appear both in columns with 'Broad Bean Origin' values and those without.

In [31]:
has_bbo_idx = df['Broad Bean Origin'].apply(lambda x: x.strip()).str.len() > 0

sbo_bbo = df[has_bbo_idx]['Specific Bean Origin']
sbo_no_bbo = df[~has_bbo_idx]['Specific Bean Origin']

pd.merge(sbo_bbo, sbo_no_bbo)

Specific Bean Origin
0       Signature Blend
1                 Blend
2                 Blend
3                 Blend
4              Amazonas
5              Amazonas
6              Amazonas
7              Amazonas
8                  Dark
9                  Dark
10                 Dark
11                 Dark
12                  Raw
13                  Raw
14                  Raw
15                  Raw
16              Orinoco

We have overlap, which is good. In theory, we could use the 'Broad Bean Origin' values from bars that *have* that value to fill in the 'Broad Bean Origin' for bars from the same specific region that *don't have* it.

But look closely at those 'Specific Bean Origin' values. Dark? Raw? Blend?

Those are specific origins. The only two origins that seem even close to regions are 'Amazonas' and 'Orinoco'. Let's look closer at the data for those regions.

In [ ]:
df[(df['Specific Bean Origin'] == 'Orinoco') | 
   (df['Specific Bean Origin'] == 'Amazonas')]

Yuck! Amazonas turns out to be a very common location. There are states called Amazonas in Brazil, Venezuela, and Peru. Orinoco is a river that runs through both Venezuela and Columbia.

In neither case do we have definitive data to make the call about the 'Broad Bean Origin' for these rows.

Unfortunately that is how it goes when working with data. You get imperfect data into your system, and then you try to research and find the best fix. But you sometimes just have to accept that you are missing data.

#### Exercise 6: Unknown Broad Bean Origins

We have a few 'Broad Bean Origin' values of `0xa0`. Change those values to the literal string 'Unknown'.

**Student Solution**

In [32]:
# Your Code Goes Here
df['Broad Bean Origin'] = np.where(df['Broad Bean Origin'] == chr(0xa0), 'Unknown', df['Broad Bean Origin'])

---

### Column: Cocoa Percent

Next we will check out the 'Cocoa Percent' column. Remember that 'Cocoa Percent' is "*Cocoa percentage (darkness) of the chocolate bar*."

As usual, we'll first see if there is any missing data:

In [33]:
df['Cocoa Percent'].isna().any()

False

Nothing missing. Great!

Next, we should probably check to make sure that the percentages fall within a valid range: 0-100 or 0.0-1.0. You might recall that 'Cocoa Percent' isn't actually a numeric column, though, so we can't easily find the range. If we sample the data, we see that it looks like percentages from 0 to 100, but they are stored as strings with '%' symbols appended.

In [34]:
df['Cocoa Percent'].sample(10)

833     82%
1689    66%
1469    73%
785     67%
1534    70%
1137    75%
893     85%
827     70%
1264    76%
1188    70%
Name: Cocoa Percent, dtype: object

We need to remove those percentage signs and convert the digits that remain into numbers. There are a few ways that we can accomplish this.

One is to apply a lambda to each value. The lambda can slice all but the last character of each value and then convert it to a float using core Python syntax.

In [35]:
df['Cocoa Percent'].apply(lambda s: float(s[:-1]))

0       63.0
1       70.0
2       70.0
3       70.0
4       70.0
        ... 
1790    70.0
1791    65.0
1792    65.0
1793    62.0
1794    65.0
Name: Cocoa Percent, Length: 1795, dtype: float64

An alternative is to use `.str.strip('%')` on the `Series` to remove the percentage sign and then pass the resultant `Series` to `pd.to_numeric()` in order to convert the string values to numbers.

In [ ]:
pd.to_numeric(df['Cocoa Percent'].str.strip('%'))

Is one way better than the other? Not necessarily. Feel free to choose whichever feels more natural to you.

Either way, we need to do the conversion and save the new values to 'Cocoa Percent'.

In [ ]:
df['Cocoa Percent'] = df['Cocoa Percent'].apply(lambda s: float(s[:-1]))
df['Cocoa Percent'].describe()

We have now converted our 'Cocoa Percent' column from a string to a floating point number. We can see in the output of the call to `describe()` that the minimum cocoa percentage that we have is 42% and that the maximum is 100%. Both seem like reasonable values for cocoa content in a chocolate bar, so our work here is done.

### Column: REF

The 'REF' column is "*A value linked to when the review was entered in the database. Higher = more recent*." Let's take a look at it.

As always, we should check and see if there are any values missing.

In [ ]:
df['REF'].isna().any()

We can `describe()` the data to see some basic statistics about it.

In [ ]:
df['REF'].describe()

Here we can see that the data ranges from 5 through 1952 and that the mean is pretty high.

Are the values unique?

In [ ]:
df['REF'].unique().size

Not unique. So 'REF' isn't a unique identifier for our rows of data.

There isn't much more that we can do with this column. We might want to visualize it to see if we can find any meaning. The numbers themselves aren't particularly interesting, but the quantity of each number might be. Let's find and plot the count of each 'REF'.

In [ ]:
import matplotlib.pyplot as plt

ref_counts = df['REF'].groupby(df['REF']).count()
plt.figure(figsize=(20,10))
plt.bar(ref_counts.index.values, ref_counts)
plt.show()

From this chart we can see that 'REF' values repeat between 1 and 9 times with 4 being the most common. Overall, there isn't much interesting data or data repair for this column.

### Column: Review Date

Review date is the date that the review for a given row was actually published. It is a numeric column.

First, let's see if any data is missing.

In [36]:
df['Review Date'].isna().any()

False

No missing data. Good.

Now we can check some basic statistics about the data.

In [ ]:
df['Review Date'].describe()

We can see publication dates ranging from 2006 through 2017, which seems like reasonable years. If we had seen dates from the 1800s or the future, we should be worried. This range seems well within reason, though.

There isn't much else that we need to do for this column. Since we only have a few years when reviews were posted, we can create a visualization showing how many reviews were posted each year.

#### Exercise 7: Reviews Per Year

Create a visualization that shows the number of reviews that were created each year.

**Student Solution**

In [49]:
import seaborn as sns
import matplotlib as plt

df_grouped_by_year = df.groupby('Review Date')['Review Date'].count()
df_grouped_by_year.columns = ['Count']
df_grouped_by_year.reset_index
df_grouped_by_year

# plt.figure(figsize=(12,6))
# plt.title('Reviews per year')
# sns.lineplot
# plt.xlabel('Year')
# plt.ylabel('')

Review Date
2006     72
2007     77
2008     93
2009    123
2010    111
2011    165
2012    195
2013    184
2014    247
2015    285
2016    219
2017     24
Name: Review Date, dtype: int64

---

### Column: Rating

We have now made it to the rating column. The rating is the "*expert rating for the bar*."  From the [documentation](https://www.kaggle.com/rtatman/chocolate-bar-ratings), the possible ratings are:

Rating | Meaning
-------|---------
5 | Elite (Transcending beyond the ordinary limits)
4 | Premium (Superior flavor development, character and style)
3 | Satisfactory (3.0) to praiseworthy(3.75) (well made with special qualities)
2 | Disappointing (Passable but contains at least one significant flaw)
1 | Unpleasant (mostly unpalatable)

Let's take a look at ratings. First off, are any missing?

In [ ]:
df['Rating'].isna().any()

Nothing missing. Let's describe the column of data.

In [ ]:
df['Rating'].describe()

It looks like our ratings are indeed floating point values and that they range from 1.0 to 5.0. But are they really continuous?

In [50]:
sorted(df['Rating'].unique())

[1.0, 1.5, 1.75, 2.0, 2.25, 2.5, 2.75, 3.0, 3.25, 3.5, 3.75, 4.0, 5.0]

Interestingly enough, the values don't seem to be continuous, but instead seem to be divided into quarters. Instead of infinite possible values between 1.0 and 5.0, we really have 17 possible values: 1.0, 1.25, 1.5, 1.75, 2.0, 2.25, 2.5, 2.75, 3.0, 3.25, 3.5, 3.75, 4.0, 4.25, 4.5, 4.75, 5.0.

What does this mean for us?

It means that downstream we might be able to use a regression or categorical model in order to predict these values.

If we think about the ratings numbers, their relative position matters. For example, a 4.0 chocolate is better than a 2.0 chocolate. But does the magnitude matter? Is a 4.0 chocolate twice as good as a 2.0 chocolate? What does that even mean?

Let's set our modelers up for success and create a new column that they can use to potentially build models for our data.

#### Exercise 8: Ratings as Categories

In this exercise we are going to create a new column called 'Grade'. Grade is a categorical rating system that maps the following ratings to grades:

Rating | Grade
-------|------
5.00   | A
4.75   | B
4.50   | C
4.25   | D
4.00   | E
3.75   | F
3.50   | G
3.25   | H
3.00   | I
2.75   | J
2.50   | K
2.25   | L
2.00   | M
1.75   | N
1.50   | O
1.25   | P
1.00   | Q

Create the 'Grade' column and add it to our chocolate bar `DataFrame`.

**Student Solution**

In [62]:
import string
rating_s = np.arange(1.00, 5.25, 0.25)

# make a series consists of 'A' to 'Q'
grade_s = pd.Series(list(string.ascii_uppercase)[:len(rating_s)])

map_rating_to_grade = dict(zip(rating_s, grade_s))
map_rating_to_grade

{1.0: 'A',
 1.25: 'B',
 1.5: 'C',
 1.75: 'D',
 2.0: 'E',
 2.25: 'F',
 2.5: 'G',
 2.75: 'H',
 3.0: 'I',
 3.25: 'J',
 3.5: 'K',
 3.75: 'L',
 4.0: 'M',
 4.25: 'N',
 4.5: 'O',
 4.75: 'P',
 5.0: 'Q'}

In [64]:
# Create nnew column called grade using map_rating_to_grade
df['Grade'] = df['Rating'].map(map_rating_to_grade)
df

Company Company Location  Bean Type Specific Bean Origin  \
0     A. Morin           France    Unknown          Agua Grande   
1     A. Morin           France    Unknown                Kpime   
2     A. Morin           France    Unknown               Atsane   
3     A. Morin           France    Unknown                Akata   
4     A. Morin           France    Unknown               Quilla   
...        ...              ...        ...                  ...   
1790    Zotter          Austria    Unknown                 Peru   
1791    Zotter          Austria  Forastero                Congo   
1792    Zotter          Austria  Forastero         Kerala State   
1793    Zotter          Austria    Unknown         Kerala State   
1794    Zotter          Austria    Unknown   Brazil, Mitzi Blue   

     Broad Bean Origin Cocoa Percent   REF  Review Date  Rating Grade  
0             Sao Tome           63%  1876         2016    3.75     L  
1                 Togo           70%  1676         2015    2.75     H  
2                 Togo           70%  1676         2015    3.00     I  
3                 Togo           70%  1680         2015    3.50     K  
4                 Peru           70%  1704         2015    3.50     K  
...                ...           ...   ...          ...     ...   ...  
1790              Peru           70%   647         2011    3.75     L  
1791             Congo           65%   749         2011    3.00     I  
1792             India           65%   749         2011    3.50     K  
1793             India           62%   781         2011    3.25     J  
1794            Brazil           65%   486         2010    3.00     I  

[1795 rows x 10 columns]

## Onward

We've now explored every column in our dataset. We have filled in missing values and repaired obviously bad data.

As you can imagine, you can spend near infinite time trying to get a dataset into shape for analysis and modeling. It is common to hear that up to 60% to 80% of a data scientist's time is spent working on the data before it is fed to a model!

In this lab, we only tried to get the data into the state that it was intended to be in. Once we get into modeling, we will learn even more data manipulation techniques that need to be used in order to get models to train well on the data.

But we aren't quite ready for model building yet. There is still more Exploratory Data Analysis (EDA) to do. In part 2 of this unit, we will look more closely at the relationships between the columns.